### PR TITLE
Refactor `RLMSyncPermissionResults` to refine `RLMResults`

### LINKED ISSUE #5075: SyncPermissionResults's method names aren't consistent with RealmCollection
## Goals

The query and sorting capability on `SyncPermissionResults` is exposed using the Objective-C API's naming conventions rather than those used in the Swift API.

## Expected Results
```swift
// Not available!
permissions.filter(predicate)
permissions.sorted(byKeyPath: keyPath, ascending: true)

// Instead:
permissions.objects(with: predicate)
permissions.sortedResults(using: keyPath, ascending: true)
```

### LINKED ISSUE #5076: (RLM)SyncPermissionResults notification does not return collection change information
## Goals
The callback passed to `(RLM)SyncPermissionResults` does not support the full collection notification capabilities and instead just returns an `NSError`:
```objc
[permissions addNotificationBlock:^(NSError *error) {
    // Permissions Changed
}];
```
This block is defined as a [`RLMPermissionStatusBlock`](https://github.com/realm/realm-cocoa/blob/dd0f2a24d8eda2241b599036eb83aff00106569c/Realm/RLMSyncUser.h#L44).

## Expected Results
I expect that the block would return collection change information so that I would know the inserted, deleted, or modified permissions as such:
```objc
[permissions addNotificationBlock:^(RLMSyncPermissionResults *permissions, 
                                         RLMCollectionChange *changes, 
                                                     NSError *error) {
    // Inserted permissions indexes
    NSArray *insertedIndexes = changes.insertions;

    // Deleted permissions indexes
    NSArray *deletedIndexes = changes.deletions;

    // Modified permissions indexes
    NSArray *modifiedIndexes = changes.modifications;
}];
```

## Actual Results
Block returns a nullable `NSError`.

### PR DESCRIPTION
**Pull request by austinzheng**
~Work in progress~

Change `RLMSyncPermissionResults` to be a private subclass of `RLMResults<RLMSyncPermissionValue *>`, and change `SyncPermissionResults` to simply be a typealias for `Results<SyncPermissionValue>`.

This reduces our API surface, adds a bunch of missing functionality, and probably fixes some latent bugs at the same time.

TODO:
- [x] Implement the rest of RLMSyncPermissionResults
- [x] Swift-specific affordances?
- [x] Migration affordances?
- [x] More tests
- [x] Compatibility with older versions of Swift
- [x] ~Filter out "no rights at all" permissions~
- [x] Object store changes merged in (https://github.com/realm/realm-object-store/pull/507)
- [x] Prerequisite Cocoa changes (`tg/relax-collection-constraints`) merged into `master-3.0`
- [x] Retarget to `master-3.0`

Fixes #5075 
Fixes #5076

### CODE REVIEW

File: Realm/RLMSyncPermissionResults.mm [hunk 1]
```diff
@@ -72,54 +102,111 @@ - (RLMSyncPermission *)lastObject {
...
+
+- (id)aggregate:(__unused NSString *)property
+         method:(__unused util::Optional<Mixed> (Results::*)(size_t))method
+     methodName:(__unused NSString *)methodName returnNilForEmpty:(__unused BOOL)returnNilForEmpty {
+    // We don't support any of the min/max/average/sum APIs; they don't make sense for this collection type.
+    return nil;
+}
+
+- (id)valueForKey:(__unused NSString *)key {
+    @throw RLMException(@"valueForKey: is not supported for RLMSyncPermissionResults");
```
Comment (bdash) [inline Realm/RLMSyncPermissionResults.mm]: Why not?

**Comment by jpsim** [comment]
Why keep `{RLM}SyncPermissionResults` around at all anymore when it's representable with existing types?

**Comment by austinzheng** [comment]
TODO: Can we use `+[RLMRealm realmWithSharedRealm:schema:` to back this `Results` internally with a Realm and hence not need to override some of the functionality in the subclass?

**Comment by austinzheng** [comment]
This doesn't have to be reviewed soon; I'm just trying out the ZenHub integration.

File: Realm/RLMResults_Private.hpp [hunk 2]
```diff
@@ -16,12 +16,36 @@
...
+// Utility functions
+
+[[gnu::noinline]]
+[[noreturn]]
+void RLMThrowResultsError(NSString * _Nullable aggregateMethod);
+
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wnullability-completeness"
+template<typename Function>
+static auto translateErrors(Function&& f, NSString *aggregateMethod=nil) {
```
Comment (tgoyne) [inline Realm/RLMResults_Private.hpp]: This needs a more specific name if it's declared in a header.

**Pull request merged**